### PR TITLE
[WOR-922] Azure clone workspace e2e test.

### PR DESCRIPTION
**Pull request by cahrens**
Ticket: https://broadworkbench.atlassian.net/browse/WOR-922

Test passed 3/3 times after explicitly specifying Hermione as the owner user:

https://github.com/broadinstitute/rawls/actions/runs/5070595939

image

---

**PR checklist**

- [ ] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should [publish a new official `rawls-model`](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model) and update `rawls-model` in [Orchestration's dependencies](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/project/Dependencies.scala).
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala [hunk 1]
```diff
@@ -43,14 +55,100 @@ class AzureWorkspacesSpec extends AnyFlatSpec with Matchers with CleanUp {
         Set.empty,
         Map("disableAutomaticAppCreation" -> "true")
       )
+      try {
+        val response = workspaceResponse(Rawls.workspaces.getWorkspaceDetails(projectName, workspaceName))
+        response.workspace.name should be(workspaceName)
+        response.workspace.cloudPlatform should be(Some(WorkspaceCloudPlatform.Azure))
+        response.workspace.workspaceType should be(Some(WorkspaceType.McWorkspace))
+      } finally {
+        Rawls.workspaces.delete(projectName, workspaceName)
```
Comment (cahrens) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala:60]: I just moved the workspace deletion into a finally clause-- select Hide Whitespace to see this.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala [hunk 2]
```diff
@@ -32,7 +41,10 @@ class AzureWorkspacesSpec extends AnyFlatSpec with Matchers with CleanUp {
     Some(UUID.fromString("f41c1a97-179b-4a18-9615-5214d79ba600"))
   )
 
+  private val wsmUrl = Rawls.url.replaceFirst("rawls", "workspace")
```
Comment (cahrens) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala:44]: This is certainly a hack, but I think I would have to do the following to get the URL without hacking that we already know about:

1. Add a workspace manager URL to the [conf template file](https://github.com/broadinstitute/firecloud-automated-testing/blob/cfea0225cc3eabb6041552f616af424530af37ee/configs/rawls/application.conf.ctmpl#L14) in `firecloud-automated-testing`.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsConfig.scala [hunk 3]
```diff
@@ -5,4 +5,6 @@ import org.broadinstitute.dsde.workbench.config.CommonConfig
 object RawlsConfig extends CommonConfig {
   private val gcs = config.getConfig("gcs")
   val pathToQAJson = gcs.getString("qaJsonFile")
+
+  val wsmUrl = config.getConfig("fireCloud").getString("workspaceManagerApiUrl")
```
Comment (cahrens) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsConfig.scala]: Depends on https://github.com/broadinstitute/terra-github-workflows/pull/78 (already merged).

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala [hunk 4]
```diff
@@ -1,29 +1,34 @@
...
 
 
 @WorkspacesAzureTest
 class AzureWorkspacesSpec extends AnyFlatSpec with Matchers with CleanUp {
-  val owner: Credentials = UserPool.chooseProjectOwner
-
-  def ownerAuthToken: AuthToken = {
-    owner.makeAuthToken()
-  }
+  val owner: Credentials = UserPool.userConfig.Owners.getUserCredential("hermione")
```
Comment (cahrens) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala:31]: `chooseProjectOwner` apparently didn't always return Hermione, which resulted in not having access to the managed app.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala [hunk 5]
```diff
@@ -32,25 +37,116 @@ class AzureWorkspacesSpec extends AnyFlatSpec with Matchers with CleanUp {
     Some(UUID.fromString("f41c1a97-179b-4a18-9615-5214d79ba600"))
   )
 
-  implicit val token: AuthToken = ownerAuthToken
+  private val wsmUrl = RawlsConfig.wsmUrl
+
+  implicit val system = ActorSystem()
 
   "Rawls" should "allow creation and deletion of azure workspaces" in {
+    implicit val token = owner.makeAuthToken()
```
Comment (cahrens) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala:45]: While trying to figure out why the test was failing when run by the workflow, I moved the implicit into each of the individual test methods. I could probably move it back out, but it also seems OK to leave like this (which is actually how it's done in other test files).

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala [hunk 6]
```diff
@@ -32,25 +37,116 @@ class AzureWorkspacesSpec extends AnyFlatSpec with Matchers with CleanUp {
...
+        response.workspace.cloudPlatform should be(Some(WorkspaceCloudPlatform.Azure))
+        response.workspace.workspaceType should be(Some(WorkspaceType.McWorkspace))
+      } finally {
+        Rawls.workspaces.delete(projectName, workspaceName)
+        assertNoAccessToWorkspace(projectName, workspaceName)
+      }
+    }
+  }
+
+  it should "allow access to WorkspaceManager API" in {
```
Comment (cahrens) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala:66]: Similar to other test files, created an explicit test point for verifying that we have access to WSM API.

**Comment by cahrens** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsConfig.scala [hunk 7]
```diff
@@ -5,4 +5,12 @@ import org.broadinstitute.dsde.workbench.config.CommonConfig
 object RawlsConfig extends CommonConfig {
   private val gcs = config.getConfig("gcs")
   val pathToQAJson = gcs.getString("qaJsonFile")
+
+  private val fireCloud = config.getConfig("fireCloud")
+  // This value is only available when tests are run through the rawls-swat Github action (in a bee).
+  val wsmUrl = if (fireCloud.hasPath("workspaceManagerApiUrl")) {
+    fireCloud.getString("workspaceManagerApiUrl")
+  } else {
+    "ERROR: workspaceManagerApiUrl unspecified! "
```
Comment (cahrens) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/RawlsConfig.scala:14]: This is how the error would present if you try to run the tests with the config option missing (such as Jenkins swatomation job):

```
ERROR: workspaceManagerApiUrl unspecified! status (of class java.lang.String)
scala.MatchError: ERROR: workspaceManagerApiUrl unspecified! status (of class java.lang.String)
	at org.broadinstitute.dsde.workbench.service.RestClient.toUri$1(RestClient.scala:42)

**Review by blakery — APPROVED** [review]

**Comment by cahrens** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala (hunk 4 above)
Comment (marctalbott) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspacesAzureApiSpec.scala:31]: afaik the only owner test user is Hermione. Did you happen to see what the other owner test user was?

**Comment by cahrens** [comment]
jenkins retest

**Review by marctalbott — APPROVED** [review]

**Comment by cahrens** [comment]
@marctalbott No, I didn't see what it was. But specifying hermione definitely fixed the problem (and another test does it as well).

**Comment by cahrens** [comment]
jenkins retest

**Pull request merged**